include/coap/encode.h

```diff
@@ -46,16 +46,41 @@ extern int coap_flsll(long long i);
 #define COAP_PSEUDOFP_ENCODE_8_4_UP(v,ls,s) (v < HIBIT ? v : (ls = coap_fls(v) - Nn, (s = (((v + ((1<<ENCODE_HEADER_SIZE<<ls)-1)) >> ls) & MMASK)), s == 0 ? HIBIT + ls + 1 : s + ls))
 
 /**
- * Decodes multiple-length byte sequences. buf points to an input byte sequence
- * of length len. Returns the decoded value.
+ * Decodes multiple-length byte sequences. @p buf points to an input byte
+ * sequence of length @p length. Returns the decoded value.
+ *
+ * @param buf The input byte sequence to decode from
+ * @param length The length of the input byte sequence
+ *
+ * @return      The decoded value
+ */
+unsigned int coap_decode_var_bytes(const uint8_t *buf, unsigned int length);
+
+/**
+ * Encodes multiple-length byte sequences. @p buf points to an output buffer of
+ * sufficient length to store the encoded bytes. @p value is the value to
+ * encode.
+ * Returns the number of bytes used to encode @p value or 0 on error.
+ *
+ * @param buf    The output buffer to decode into
+ * @param length The output buffer size to encode into (must be sufficient)
+ * @param value  The value to encode into the buffer
+ *
+ * @return       The number of bytes used to encode @p value or @c 0 on error.
  */
-unsigned int coap_decode_var_bytes(const uint8_t *buf, unsigned int len);
+unsigned int coap_encode_var_safe(uint8_t *buf,
+                                  unsigned int length,
+                                  unsigned int value);
 
 /**
- * Encodes multiple-length byte sequences. buf points to an output buffer of
- * sufficient length to store the encoded bytes. val is the value to encode.
- * Returns the number of bytes used to encode val or 0 on error.
+ * @deprecated.  Provided for backward compatability.  As @p value is only of
+ * length sizeof(unsigned int), and buf usually is defined as an array, it
+ * is unsafe to continue to use this variant if buf is too small.
```

Comment:
This makes no sense because by definition, the function must only be called with a buffer that is large enough for the representation of val.

Reply:
Old Code in examples/client.c has for example this
static coap_tid_t
clear_obs(coap_context_t *ctx, coap_session_t *session) {
  coap_pdu_t *pdu;
  coap_optlist_t *option;
  coap_tid_t tid = COAP_INVALID_TID;
  unsigned char buf[2];
  ...
    if (!coap_add_option(pdu,
      COAP_OPTION_OBSERVE,
      coap_encode_var_bytes(buf, COAP_OBSERVE_CANCEL),
      buf)) {
    coap_log(LOG_CRIT, "cannot add option Observe: %u\n", COAP_OBSERVE_CANCEL);
    goto error;
  }

if COAP_OBSERVE_CANCEL happened to be bigger than 2 bytes, buf[2] would over flow.  Test changed to
\- * is unsafe to continue to use this variant if buf is too small.
\+ * is unsafe to continue to use this variant if buf[] is less than buf[4].